### PR TITLE
pgsql: track 'progress' in tx per direction - v4

### PR DESCRIPTION
**Pull request by jufajardini**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/
https://redmine.openinfosecfoundation.org/issues/7113

Previous PR: https://github.com/OISF/suricata/pull/11618

Describe changes:
- track `tx_completion` per tx direction
- add more `PgsqlStateProgress` states to tx completion check (especially for `to_server` direction)
- reorganize `PgsqlStateProgress` enum to separate `to_server` and `to_client` states
- bring the trigger raw stream reassembly call to the moment when tx is completed

Update:
- add comment explaining and referencing Pgsql state machine
- make code a bit more compact when checking for temporary progress state in `parse_request`
- try to address https://github.com/OISF/suricata/pull/11618#discussion_r1708653138 by not checking for `TxReqDone` before flushing out tx
- update all pgsql function exposed to C to the new API style

Provide values to any of the below to override the defaults.

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2014

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 22141

File: rust/src/pgsql/pgsql.rs [hunk 1]
```diff
@@ -36,17 +36,27 @@ static mut PGSQL_MAX_TX: usize = 1024;
...
+    TxReqFlushedOut,
+}
+
+#[repr(u8)]
+#[derive(Copy, Clone, PartialOrd, PartialEq, Eq, Debug)]
+pub enum PgsqlTxResProgress {
+    TxResInit = 0,
+    TxResReceived,
+    TxResDone,
+    TxResFlushedOut,
```
Comment (catenacyber) [inline rust/src/pgsql/pgsql.rs:52]: Values look the same for req and res : could it be one enum ? used by two fields

File: rust/src/pgsql/pgsql.rs [hunk 2]
```diff
@@ -343,14 +360,28 @@ impl PgsqlState {
...
                     if let Some(tx) = self.find_or_create_tx() {
                         tx.request = Some(request);
                         if tx_completed {
-                            tx.tx_state = PgsqlTransactionState::ResponseDone;
+                            if temp_state == PgsqlStateProgress::ConnectionTerminated || temp_state == PgsqlStateProgress::CancelRequestReceived { 
+                                tx.tx_req_state = PgsqlTxReqProgress::TxReqDone;
+                                /* The server won't send any responses to such requests, so transaction should be over */
+                                tx.tx_res_state = PgsqlTxResProgress::TxResDone;
+                            } else {
+                                tx.tx_req_state = PgsqlTxReqProgress::TxReqDone;
```
Comment (catenacyber) [inline rust/src/pgsql/pgsql.rs:382]: Done is every case

File: rust/src/pgsql/pgsql.rs [hunk 3]
```diff
@@ -343,14 +360,28 @@ impl PgsqlState {
...
                         self.state_progress = state;
+                        // PostreSQL progress states can be represented as a finite state machine
+                        // After the connection phase, the backend/ server will be mostly waiting in a state of `ReadyForQuery`, unless it's processing some request.
+                        // When the frontend wants to cancel a request, it will send a CancelRequest message over a new connection - to which there won't be any responses.
+                        // If the frontend wants to terminate the connection, the backend won't send any confirmation after receiving a Terminate request.
+                        // A simplified finite state machine for PostgreSQL v3 can be found at:
+                        // https://samadhiweb.com/blog/2013.04.28.graphviz.postgresv3.html
+                        (state, Self::request_is_complete(state))
+                    } else {
+                        (self.state_progress, false)
```
Comment (catenacyber) [inline rust/src/pgsql/pgsql.rs:372]: Thanks.

This is complex stuff and needs even more doc/polishing I guess :
Could you rename `state` to tell if it is a Tx progress/state or a PgsqlState progress ?

Could tx_completed be an option of some state ?

File: rust/src/pgsql/pgsql.rs [hunk 4]
```diff
@@ -97,28 +108,30 @@ impl PgsqlTransaction {
...
+    SASLAuthenticationReceived,
     SASLAuthenticationContinueReceived,
     SASLAuthenticationFinalReceived,
     SimpleAuthenticationReceived,
-    PasswordMessageReceived,
     AuthenticationOkReceived,
     ParameterSetup,
     BackendKeyReceived,
     ReadyForQueryReceived,
-    SimpleQueryReceived,
```
Comment (catenacyber) [inline rust/src/pgsql/pgsql.rs:115]: Could the reorder be its own commit ?
(to keep the complex commit as simple as possible)

File: rust/src/pgsql/pgsql.rs [hunk 5]
```diff
@@ -706,19 +762,33 @@ pub extern "C" fn rs_pgsql_state_get_tx_count(state: *mut std::os::raw::c_void)
 }
 
 #[no_mangle]
-pub extern "C" fn rs_pgsql_tx_get_state(tx: *mut std::os::raw::c_void) -> PgsqlTransactionState {
+pub extern "C" fn rs_pgsql_tx_get_req_state(tx: *mut std::os::raw::c_void) -> PgsqlTxReqProgress {
```
Comment (catenacyber) [inline rust/src/pgsql/pgsql.rs]: This function is pure rust, no extern C...

**Review by catenacyber — CHANGES_REQUESTED** [review]
Looks nice, some nits.
The challenge is to make this complex machine state the most readable possible

File: rust/src/pgsql/pgsql.rs (hunk 1 above)
Comment (jufajardini) [reply rust/src/pgsql/pgsql.rs:52]: Likely. I had understood that Victor preferred them separated, maybe he'll see this differently when he returns...

**Comment by jufajardini** [comment]
> Looks nice, some nits. The challenge is to make this complex machine state the most readable possible

And there's more to be added, with some sub-protocols... But hopefully this would be more self-contained...

File: rust/src/pgsql/pgsql.rs (hunk 5 above)
Comment (jufajardini) [reply rust/src/pgsql/pgsql.rs]: good catch, sir 🙇🏽

File: rust/src/pgsql/pgsql.rs (hunk 1 above)
Comment (jufajardini) [reply rust/src/pgsql/pgsql.rs:52]: It does feel nicer to see the enum variant and be able to tell whether it's a request or response, but... ok, that's probably not reason enough.

File: rust/src/pgsql/pgsql.rs [hunk 6]
```diff
@@ -341,16 +358,30 @@ impl PgsqlState {
             );
             match PgsqlState::state_based_req_parsing(self.state_progress, start) {
                 Ok((rem, request)) => {
-                    sc_app_layer_parser_trigger_raw_stream_reassembly(flow, Direction::ToServer as i32);
                     start = rem;
-                    if let Some(state) = PgsqlState::request_next_state(&request) {
+                    let (temp_state, tx_completed) = if let Some(state) = PgsqlState::request_next_state(&request) {
```
Comment (victorjulien) [inline rust/src/pgsql/pgsql.rs:362]: what is temp about the state?

Could it just be `let (state, tx_completed) = if let Some(state) = ....`?

Comment (catenacyber) [reply rust/src/pgsql/pgsql.rs:362]: Pressing https://github.com/OISF/suricata/pull/11635/files/057c398f669c5b174d555116c87c8d1689db98cf#r1734188284
```rust
let new_state = PgsqlState::request_next_state(&request);
if let Some(state) = new_state {
    self.state_progress = state;
}
...
if let Some(tx) = self.find_or_create_tx() {
    tx.request = Some(request);
    if let Some(state) = new_state {
        if  Self::request_is_complete(state) && ...
```
